### PR TITLE
dag/expr: implement log2 && log10

### PR DESCRIPTION
**Pull request by sllt**
## What have you changed? 

* add `log2` && `log10` functions in `builtin_math.rs`

## What are the type of the changes? 

Improvement

## How has this PR been tested? 

Unit test

## Does this PR affect documentation (docs/docs-cn) update? 

No

## Does this PR affect tidb-ansible update?

No

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 1]
```diff
@@ -298,4 +310,44 @@ mod test {
...
+
+    #[test]
+    fn test_log2() {
+        let tests = vec![
+            (ScalarFuncSig::Log2, Datum::F64(16f64), Datum::F64(4f64)),
+            (
+                ScalarFuncSig::Log2,
+                Datum::F64(5f64),
+                Datum::F64(2.321928094887362f64),
+            ),
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_math.rs]: how about case log2(-0.123)?

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 2]
```diff
@@ -298,4 +310,44 @@ mod test {
...
+
+    #[test]
+    fn test_log10() {
+        let tests = vec![
+            (ScalarFuncSig::Log10, Datum::F64(100f64), Datum::F64(2f64)),
+            (
+                ScalarFuncSig::Log10,
+                Datum::F64(101f64),
+                Datum::F64(2.0043213737826426f64),
+            ),
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_math.rs]: How about case log10(-0.21243)?

**Comment by sllt** [comment]
@AndreMouche merged master & fiexd conflicts

**Comment by AndreMouche** [comment]
@breeswish PTAL

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 3]
```diff
@@ -308,6 +330,50 @@ mod test {
...
+    }
+
+    #[test]
+    fn test_log10() {
+        let tests = vec![
+            (ScalarFuncSig::Log10, Datum::F64(100f64), Datum::F64(2f64)),
+            (
+                ScalarFuncSig::Log10,
+                Datum::F64(101f64),
+                Datum::F64(2.0043213737826426f64),
```
Comment (liufuyang) [inline src/coprocessor/dag/expr/builtin_math.rs]: @sllt  We had some discussion of whether to use a specific f64 number, as different architect might give slight different result, making the test fails.

You can see more discussion here: https://github.com/tikv/tikv/pull/3406

I think it is safe to just use `assert!((got.f64() - exp).abs() < 1e-10_f64);` to do comparison, instead of doing the strong == on f64.

You can see the official Rust f64 doc is doing the same:
https://doc.rust-lang.org/std/primitive.f64.html#method.log2

**Comment by AndreMouche** [comment]
Please fix the conflicts

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 4]
```diff
@@ -109,6 +109,28 @@ impl ScalarFunc {
         self.children[0].eval_int(ctx, row)
     }
 
+    #[inline]
+    pub fn log2(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        let n = try_opt!(self.children[0].eval_real(ctx, row));
+        let res = n.log2();
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_math.rs]: What if n is negative?

**Comment by huachaohuang** [comment]
@BusyJay @breeswish @sllt PTAL

File: src/coprocessor/dag/expr/builtin_math.rs (hunk 4 above)
Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_math.rs]: So `res.is_finite()` will return false.

File: src/coprocessor/dag/expr/builtin_math.rs (hunk 3 above)
Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_math.rs]: Any updates?

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 5]
```diff
@@ -347,6 +369,50 @@ mod test {
...
 
+    #[test]
+    fn test_log2() {
+        let tests = vec![
+            (ScalarFuncSig::Log2, Datum::F64(16f64), Datum::F64(4f64)),
+            (
+                ScalarFuncSig::Log2,
+                Datum::F64(5f64),
+                Datum::F64(2.321928094887362f64),
+            ),
```
Comment (winoros) [inline src/coprocessor/dag/expr/builtin_math.rs]: Add a case for `log2(null)`?

Comment (sllt) [reply src/coprocessor/dag/expr/builtin_math.rs]: added

File: src/coprocessor/dag/expr/builtin_math.rs (hunk 4 above)
Comment (sllt) [reply src/coprocessor/dag/expr/builtin_math.rs]: will return null

**Comment by huachaohuang** [comment]
@winoros @hicqu PTAL

**Comment by sllt** [comment]
@breeswish  @huachaohuang PTAL

File: src/coprocessor/dag/expr/scalar_function.rs [hunk 6]
```diff
@@ -1159,6 +1161,9 @@ mod test {
                     ScalarFuncSig::JsonUnquoteSig,
                     ScalarFuncSig::ASCII,
                     ScalarFuncSig::Bin,
+                    ScalarFuncSig::BitNegSig,
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/scalar_function.rs]: Why is this line added? Rest LGTM.

Comment (sllt) [reply src/coprocessor/dag/expr/scalar_function.rs]: @huachaohuang  sorry, the line is deleted.
